clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/AbstractGoogleServerGroupCachingAgent.java

```diff
@@ -768,25 +767,30 @@ private void setSourceImage(
     if (appversion == null) {
       return null;
     }
-    buildInfo
-        .put("package_name", appversion.getPackageName())
-        .put("version", appversion.getVersion())
-        .put("commit", appversion.getCommit());
+    ImmutableMap.Builder<String, Object> buildInfo = ImmutableMap.builder();
+    putIfNotNull(buildInfo, "package_name", appversion.getPackageName());
+    putIfNotNull(buildInfo, "version", appversion.getVersion());
+    putIfNotNull(buildInfo, "commit", appversion.getCommit());
     if (appversion.getBuildJobName() != null) {
-      Map<String, String> jenkinsInfo = new HashMap<>();
-      jenkinsInfo.put("name", appversion.getBuildJobName());
-      jenkinsInfo.put("number", appversion.getBuildNumber());
+      ImmutableMap.Builder<String, String> jenkinsInfo = ImmutableMap.builder();
+      putIfNotNull(jenkinsInfo, "name", appversion.getBuildJobName());
+      putIfNotNull(jenkinsInfo, "number", appversion.getBuildNumber());
       if (tags.containsKey("build_host")) {
         jenkinsInfo.put("host", tags.get("build_host"));
       }
-      buildInfo.put("jenkins", jenkinsInfo);
+      buildInfo.put("jenkins", jenkinsInfo.build());
```

Comment:
Should we do the same for `build_host` and `build_info_url` to guard against the `tags` (mutable) map containing these keys but mapped to `null`?  (And also just to make the logic consistent, as right now it's not clear why these two use different logic.)

Reply:
Yeah, okay, good call. I switched to a slightly different approach, building into a `HashMap` and then copying to an `ImmutableMap` at the end. It's still annoying but better, I think.

Thanks...

Reply:
:+1: new approach looks good!